### PR TITLE
feat: support SDM Subscription and Unsubscription for UE Session

### PR DESCRIPTION
**Pull request by saileshvvr**
### Description
- Add support for SDM Subscription during UE 1st PDU Session Establishment
- Add support for SDM UnSubscription during release of last UE PDU Session

### CODE REVIEW

**Comment by ianchen0119** [comment]
@andy89923 
Please help to test the functionality (w/ OAuth and w/o OAuth)

File: internal/sbi/consumer/udm_service.go [hunk 1]
```diff
@@ -199,3 +199,120 @@ func (s *nudmService) GetSmData(ctx context.Context, supi string,
...
+			sdmUri := util.SearchNFServiceUri(smf_context.GetSelf().UDMProfile, models.ServiceName_NUDM_SDM,
+				models.NfServiceStatus_REGISTERED)
+			if sdmUri == "" {
+				return nil, errors.Errorf("SMF can not select an UDM by NRF: SearchNFServiceUri failed")
+			}
+
+			SDMConf := Nudm_SubscriberDataManagement.NewConfiguration()
+			SDMConf.SetBasePath(sdmUri)
+
+			client = Nudm_SubscriberDataManagement.NewAPIClient(SDMConf)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:218]: `getSubscribeDataManagementClient()`

File: internal/sbi/consumer/udm_service.go [hunk 2]
```diff
@@ -199,3 +199,120 @@ func (s *nudmService) GetSmData(ctx context.Context, supi string,
...
+				sdmUri := util.SearchNFServiceUri(smf_context.GetSelf().UDMProfile, models.ServiceName_NUDM_SDM,
+					models.NfServiceStatus_REGISTERED)
+				if sdmUri == "" {
+					return nil, errors.Errorf("SMF can not select an UDM by NRF: SearchNFServiceUri failed")
+				}
+
+				SDMConf := Nudm_SubscriberDataManagement.NewConfiguration()
+				SDMConf.SetBasePath(sdmUri)
+
+				client = Nudm_SubscriberDataManagement.NewAPIClient(SDMConf)
```
Comment (andy89923) [inline internal/sbi/consumer/udm_service.go:281]: `getSubscribeDataManagementClient()`

File: internal/sbi/consumer/udm_service.go (hunk 1 above)
Comment (saileshvvr) [reply internal/sbi/consumer/udm_service.go]: done!

File: internal/sbi/consumer/udm_service.go (hunk 2 above)
Comment (saileshvvr) [reply internal/sbi/consumer/udm_service.go]: done!

**Comment by saileshvvr** [comment]
Hi @andy89923 , 
Addressed the comment. I did not follow the refactored code and used the legacy way.
Updated as suggested :) 
Thanks!

**Comment by saileshvvr** [comment]
Hi @andy89923 
Can you please share if in case anything missing.
As this PR functionality is pending from quite sometime, would like to check once!
Thanks

**Review by andy89923 — APPROVED** [review]
@saileshvvr Sorry for the late review! 

After testing w/wo OAuth, this PR can subscribe when a new UE and unsubscribe when the UE's last PDU session is released.  

Thanks for the PR, LGTM!

File: internal/sbi/processor/pdu_session.go [hunk 3]
```diff
@@ -121,6 +121,17 @@ func (p *Processor) HandlePDUSessionSMContextCreate(
 		}
 	}
 
+	if !p.Context().Ues.UeExists(smContext.Supi) {
```
Comment (ianchen0119) [inline internal/sbi/processor/pdu_session.go:124]: What if the error happens after the SDM data changed?
I think use the `defer` would be better in this case:
```go
defer func(err error) {
  if err != nil {
	if !p.Context().Ues.UeExists(smContext.Supi) {
		if problemDetails, err := p.Consumer().
			Subscribe(ctx, smContext, smPlmnID); problemDetails != nil {
			smContext.Log.Errorln("SDM Subscription Failed Problem:", problemDetails)
		} else if err != nil {
			smContext.Log.Errorln("SDM Subscription Error:", err)
		}
	} else {
		p.Context().Ues.IncrementPduSessionCount(smContext.Supi)
	}
  }
}(err)
```

File: internal/sbi/processor/pdu_session.go [hunk 4]
```diff
@@ -878,6 +898,15 @@ func (p *Processor) HandlePDUSessionSMContextRelease(
 		}
 	}
 
+	if p.Context().Ues.UeExists(smContext.Supi) {
```
Comment (ianchen0119) [inline internal/sbi/processor/pdu_session.go:901]: Same as above.
Please do this action by using `defer`.

File: internal/sbi/processor/pdu_session.go [hunk 5]
```diff
@@ -969,6 +998,15 @@ func (p *Processor) HandlePDUSessionSMContextLocalRelease(
...
```
Comment (ianchen0119) [inline internal/sbi/processor/pdu_session.go:1001]: Same as above.
Please do this action by using `defer`.

File: internal/sbi/processor/pdu_session.go [hunk 6]
```diff
@@ -325,6 +336,15 @@ func (p *Processor) HandlePDUSessionSMContextUpdate(
 				}
 			}
 
+			if smf_context.GetSelf().Ues.UeExists(smContext.Supi) {
```
Comment (ianchen0119) [inline internal/sbi/processor/pdu_session.go:339]: Same as above.
Please do this action by using `defer`.

**Review by ianchen0119 — CHANGES_REQUESTED** [review]

**Comment by ianchen0119** [comment]
It looks good to me except the requested changes.

Hi @saileshvvr 
Please help to update & test the PR based on my comments.
Many thanks!

**Comment by saileshvvr** [comment]
Hi @ianchen0119 ,

Thank you for the feedback. As I was occupied with other work, missed the comments.
I shall check and update accordingly.

Thanks!

**Comment by andy89923** [comment]
Hi @saileshvvr ,

We can't use the defer function and pass the error to differentiate if the error happens after the defer function. 
Should creat an error instead.
```go
func main() {
	var err error
	defer func() {
		if err != nil {
			fmt.Printf("defer: %s\n", err)
		} else {
			fmt.Println("defer: defer not error")
		}
	}()

	err = errors.New("new error 1")
	if err != nil {
		return
	}
}

/*output
defer: new error 1
*/
```
You can refer to https://stackoverflow.com/questions/42703707/when-defer-func-evaluates-its-parameters

**Comment by saileshvvr** [comment]
Hi @andy89923 , @ianchen0119 ,
Agree with your comments. Can we handle this as a seperate improvement PR once after delivering this implementation.
After submitting this PR I shall raise a new improvement PR with the recommended changes if you are OKAY.
please share your comments. Thanks!

**Comment by andy89923** [comment]
Hi @saileshvvr ,

> Can we handle this as a separate improvement PR? 

No, we should revise this in the same PR. 

If you don't have time, I will help revise if the PR isn't ready until next Wednesday.

**Comment by saileshvvr** [comment]
Thank you @andy89923 . 
As I shared earlier, I was occupied with other tasks and may be getting time during next week.
Thanks for your quick response and support.

**Comment by andy89923** [comment]
Hi @saileshvvr ,

Sorry, I don't have time to help with this PR; please fix the comments when you are free!

**Comment by saileshvvr** [comment]
Hi @ianchen0119 , @andy89923 ,
As I understand we can keep the code with in defer{}() only during Subscribe(). 

For all other cases (Session Release(UE/NW triggered) or Local Release or Failure cases) we can just call Unsubscribe(). As Unsubscribe() will decrement the session count if more than 1 session else session HTTP DELETE to Unsubscribe.
So no change may be needed during Unsubscribe() call.

Please let me know your comments so as to update the latest patch accordingly.

Thanks
Sailesh

**Comment by saileshvvr** [comment]
Thanks @ianchen0119 @andy89923

**Review by ianchen0119 — APPROVED** [review]

**Pull request merged**